src/lib.rs

```diff
@@ -73,18 +73,43 @@ pub fn compile(cmds: Vec<Cmd>) -> Result<Program, ()> {
             _ => (),
         }
     }
-
-    Ok(Program {
-        code: cmds,
-        bg,
-        fg,
-        khz,
-    })
+    // Validate the bytebeat by checking that the stack does not get popped when empty
+    let cmds = {
+        let mut stack_size = 0 as i32;
+        for cmd in cmds.iter() {
+            stack_size += match *cmd {
+                Var | NumF(_) | NumI(_) => 1,
+                Fg(_) | Bg(_) | Khz(_) | Comment(_) => continue,
+                // These all pop 1 value off the stack and push 1
+                // value back on, so the net effect is no stack change
+                Sin | Cos | Tan => 0,
+                // Arr(x) pops a value off the stack (called the index)
+                // then pops x more values off the stack. Finally, it
+                // pushes one value back onto the stack based on the index
+                // Thus the net effect of Arr is to reduce the stack size by x.
+                Arr(x) => -(x as i32),
+                Cond => -2,
+                // Split these into multiple branches to make rustfmt stop complaining
+                Add | Sub | Mul | Div | Mod => -1,
+                Shl | Shr | And | Orr | Xor => -1,
+                Pow | AddF | SubF | MulF | DivF | ModF => -1,
+                Lt | Gt | Leq | Geq | Eq | Neq => -1,
+            };
+            if stack_size <= 0 {
+                return Err("Invalid bytebeat");
+            }
+        }
+        Ok(cmds)
+    };
+    match cmds {
```

Comment:
`Err(…)` will never get here, since you're early-returning, so you don't need the `match` here.